### PR TITLE
Light IO entry

### LINKED ISSUE #87: "Light" IO entry
rfc basically

* libinput (perhaps use [this](https://docs.rs/input/0.2.0/input/struct.Libinput.html) crate - interface is ~~~far from mature~~~ *horrible* but it's probably better than creating our own bindings) to read events from configured usb keyboard and mouse
* use the evdev ioctl to exclusively grab them
* if ga is down, do light entry instead of just refusing to do anything
* config option to do light entry by default even if ga is up
* can then bind a hotkey to upgrade to full entry
* during light entry we ignore hotkeys from ga - catch them in the driver instead
* if ga dies, downgrade from full to light entry
* on io exit, either detach usb or stop listening to libusb (but don't *downgrade* because why would you ever)
* forward mouse and all remaining keystrokes right through with qapi's `input-send-event` command (keys are obvious and libinput's `POINTER_MOTION` maps straight to relative mouse movement - might perhaps have to rescale to qemu's `0 .. 0x7ffff` range though)
* (and of course funnel the whole thing through our tokio eventloop)

# motivation

* can exit vm without ga, need this for:
  - nogui setup
  - windows updates
  - bitlocker password
  - installing / updating ga
* faster entry, ideally makes switching as seamless as alt+tab or whatever your window manager's hotkeys are

# drawbacks

* if you really definitely just want USB passthrough you have to press one more hotkey to enter - can however address this by still providing the old functionality (usb passthrough or nothing, depending on ga state) and assigning a new control command id to the new and shiny version
* need to test this but there's obviously a lot more latency - you might not notice right away and accidentally use light entry, then be disappointed about bad usb passthrough perf
* probably not too light on cpu either
* need to parse and understand windows hotkeys in the driver - good anyways so we can catch this configuration error early and not only at a time when the vm has already booted
* do we need a fully-blown qapi rpc impl for this? is there any point in waiting for the return message?

### LINKED ISSUE #111: Semi-Temporarily switch the input to windows
If you have BitLocker enabled, you obviously need to switch over your keyboard to Windows to enter your password. 

During the boot of Windows both Windows and Linux are blocked. This is usually not a problem, as the Windows boot is less than 10 seconds for me, and waiting isn't a big problem. However, Windows Update can prolongen this time where I can neither use windows nor linux significantly. 

This is the reason I usually use a second hotkey, executing this script when booting:

```bash
#!/bin/bash
echo -e -n '\x03' | netcat -w 1 -U /run/user/1000/windows-gaming-driver/control.sock
sleep 20
echo -e -n '\x04' | netcat -w 1 -U /run/user/1000/windows-gaming-driver/control.sock
```

This script exits the IO-devices again after 20 seconds, so if Windows Update takes longer, I can use Linux again. 

However, I'd like to propose an aditional command besides `IoExit`: `SoftIoExit` - it only exits when the guest agent isn't active at the moment - similar to `IoEntry` and `ForceIoEntry`, but vice versa. This has some advantages:

- The way I hand over my IO-devices is consitent, I'd only have one hotkey instead of two. If the GA isn't running for some reason, the controls will be handed back to linux after a while.
- If I boot fast there is no unnecessary entry / exit. 
- I can't make a mistake when pressing the hotkey.

### PR DESCRIPTION
**Pull request by main--**
# Blockers

## Trivial (up for grabs)

- [x] verify ioctl result
- [x] fix style (as in replace `sweg(…)` with `libc::sweg(…)` AND NAMING)
- [ ] libinput and hotkeys should not be `pub mod`
- [x] move closure thing into hotkey mod
- [x] move they keycode resolver to its own file/module
- [x] Support media keys
- [x] decide on absolute vs relative mouse events
    (this one's not really trivial but we have to make a decision either way - for now all we support is relative as that's just how a mouse works but this probably breaks touchscreen-style input devices? then again, is that really an issue? how would that work anyways? Confusing! Unless someone comes up with a better idea, I'd suggest we leave it the way it is until it breaks for someone.)

## Easy (still up for grabs but you need to actually read the code you are editing to do this)

- [x] actually implement the entry-exit state machine; call resume and suspend for that (and instantly suspend on startup)
- [x] add remaining events (mouse click; keyboard keys)
- [x] hotkey handling
- [x] reopen libinput (as they keep udev handles)
- [x] not hardcode path to my mouse - figure out via udev (filter keyboard and mouse yadda yadda)
- [x] fix usb controller handling (reserve xhci port) (or just connect a PS2 mouse)
- [x] Fix host key hang issue (~~~appears to be~~~ is a [libinput bug](https://bugs.freedesktop.org/show_bug.cgi?id=101796))
- [x] Fix travis: There is no recent libinput for the ubuntu versions supported by travis. Moved to arch-travis.

# Postponeables (can just merge and create issues for these)

* have setup wizard create a udev rule to adjust permissions on the input devnode as well (before we only needed usb devnode)
* Add some logging
* Fix the remaining host key hang issue (once that libinput patch is released, they keys are no longer stuck permanently - temporarily stuck keys are still problematic however)
* Desktop notification / Sound for light entry to notify user of fallback.

Fixes #87, fixes #111.

**Remember to add libinput AUR dependency when merging this!**

### CODE REVIEW

File: windows-gaming-driver/src/driver/libinput.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+use input::event::pointer::PointerEvent;
+use libc::{open, close, ioctl, c_char, c_int, c_ulong, c_void};
+use mio::Ready;
+
+use super::my_io::MyIo;
+
+const EVIOCGRAB: c_ulong = 1074021776;
+
+unsafe extern "C" fn do_open(x: *const c_char, y: c_int, z: *mut c_void) -> c_int {
+    let fd = open(x, y);
```
Comment (oberien) [inline windows-gaming-driver/src/driver/libinput.rs]: Why not use `fs::open` and `AsRawFd`? We'll need to `mem::forget` the `File` and use `FromRawFd` in `do_close` to properly drop the file. Still, this is definitely more rusty.

File: windows-gaming-driver/src/driver/libinput.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    pub fn suspend(&mut self) {
+        self.li.suspend();
+    }
+
+    pub fn resume(&mut self) {
+        self.li.resume().unwrap();
+    }
+}
+
+pub struct InputListener<'a>(pub &'a RefCell<Input>);
```
Comment (oberien) [inline windows-gaming-driver/src/driver/libinput.rs]: Does the internal value actually need to be public?

File: windows-gaming-driver/src/driver/monitor/codec.rs [hunk 3]
```diff
@@ -19,6 +19,19 @@ pub enum QmpCommand {
...
+    InputSendEvent {
+        events: Vec<InputEvent>,
+    },
+}
+
+#[derive(Serialize)]
+#[serde(tag = "type", content = "data", rename_all = "snake_case")]
+pub enum InputEvent {
+    Rel {
+        axis: &'static str,
```
Comment (oberien) [inline windows-gaming-driver/src/driver/monitor/codec.rs]: Make an enum out of it.

**Review by oberien — CHANGES_REQUESTED** [review]
I mean, this needs hella fixing before it's able to be merged.

File: windows-gaming-driver/src/driver/libinput.rs (hunk 2 above)
Comment (main--) [reply windows-gaming-driver/src/driver/libinput.rs]: sure or else you can't construct it

File: windows-gaming-driver/src/driver/monitor/codec.rs (hunk 3 above)
Comment (main--) [reply windows-gaming-driver/src/driver/monitor/codec.rs]: if only we had a tool that could autogenerate these definitions :thinking:

File: windows-gaming-driver/src/driver/libinput.rs (hunk 1 above)
Comment (main--) [reply windows-gaming-driver/src/driver/libinput.rs]: I disagree

what we really want here is a raw file descriptor

we are already in C land, there's no point in trying to act like this is rust code

File: windows-gaming-driver/src/driver/monitor/codec.rs (hunk 3 above)
Comment (oberien) [reply windows-gaming-driver/src/driver/monitor/codec.rs]: That'd be pretty damn fancy, that's true :thonkin:

File: windows-gaming-driver/src/driver/mod.rs [hunk 4]
```diff
@@ -67,32 +73,81 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
...
+                    InputButton::WheelDown
+                } else {
+                    InputButton::WheelUp
+                };
+
+                QmpCommand::InputSendEvent {
+                    events: iter::repeat(direction).flat_map(|b| vec![
+                        InputEvent::Btn { down: true, button: b },
+                        InputEvent::Btn { down: false, button: b },
+                    ]).take(steps.abs() as usize).collect(),
```
Comment (main--) [inline windows-gaming-driver/src/driver/mod.rs]: bug

**Comment by main--** [comment]
fwiw there's a [patch](https://lists.freedesktop.org/archives/wayland-devel/2017-July/034500.html) for libinput now

File: windows-gaming-driver/src/config.rs [hunk 5]
```diff
@@ -229,12 +229,19 @@ pub struct MachineConfig {
...
+            action: HotKeyAction::Action(Action::IoExit),
+        },
+        HotKey {
+            key: KeyBinding::new(vec![Modifier::Win], Key::Insert, true),
+            action: HotKeyAction::Action(Action::IoUpgrade),
+        },
+    ]
 }
 
+
```
Comment (oberien) [inline windows-gaming-driver/src/config.rs]: ¿

Comment (oberien) [reply windows-gaming-driver/src/config.rs]: ¿ still

File: windows-gaming-driver/src/driver/controller.rs [hunk 6]
```diff
@@ -32,13 +35,23 @@ enum State {
     Resuming,
 }
 
+#[derive(PartialEq, Eq, Clone, Copy)]
+enum IoState {
+    Detached,
+    LightEntry,
+    LightFallback,
```
Comment (moritzuehling) [inline windows-gaming-driver/src/driver/controller.rs]: Consider renaming this to something `LightAwaitingUpgrade` to better describe what it does.

File: windows-gaming-driver/src/driver/controller.rs [hunk 7]
```diff
@@ -158,26 +178,51 @@ impl Controller {
...
+            State::Resuming | State::Suspending => (),
             State::Suspended => {
                 // make them wake up
                 (&self.monitor).send(QmpCommand::SystemWakeup).unwrap();
                 // can't enter now - gotta wait for GA to get ready
                 self.ga = State::Resuming;
             },
-            State::Up | State::Pinging => self.io_force_attach()
+            State::Down => {
+                self.light_attach();
```
Comment (moritzuehling) [inline windows-gaming-driver/src/driver/controller.rs:189]: A notification for the user seems to be approperiate here. 

It is a fallback, so the user ought to know about it

File: windows-gaming-driver/src/driver/controller.rs [hunk 8]
```diff
@@ -223,14 +269,19 @@ impl Controller {
...
-            .filter(|&(_, dev)| !dev.permanent).map(|(i, _)| i) {
-            (&self.monitor).send(QmpCommand::DeviceDel { id: format!("usb{}", i) }).unwrap();
+
+        match self.io_state {
+            IoState::Detached => (),
+            IoState::LightFallback | IoState::LightEntry => self.input.borrow_mut().suspend(),
+            IoState::FullEntry => {
+                for i in self.machine_config.usb_devices.iter().enumerate()
+                    .filter(|&(_, dev)| !dev.permanent).map(|(i, _)| i) {
+                        (&self.monitor).send(QmpCommand::DeviceDel { id: format!("usb{}", i) }).unwrap();
```
Comment (oberien) [inline windows-gaming-driver/src/driver/controller.rs]: Weird indentation, would have done it the other way around (indent .filter more).

File: windows-gaming-driver/src/driver/hotkeys/keys.rs [hunk 9]
```diff
@@ -0,0 +1,209 @@
+// adapted from mono's System.Windows.Forms/Keys.cs
+#[repr(u32)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub enum Keys {
```
Comment (moritzuehling) [inline windows-gaming-driver/src/driver/hotkeys/keys.rs:4]: Should be "Kies"

File: windows-gaming-driver/src/driver/hotkeys/linux.rs [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+// const KEY_HANGUEL: u32 = KEY_HANGEUL;
+// const KEY_HANJA: u32 = 123;
+// const KEY_YEN: u32 = 124;
+const KEY_LEFTMETA: u32 = 125;
+const KEY_RIGHTMETA: u32 = 126;
+const KEY_COMPOSE: u32 = 127;
+const KEY_MEDIANEXTTRACK: u32 = 163;
+const KEY_MEDIAPLAYPAUSE: u32 = 164;
+const KEY_MEDIAPREVIOUSTRACK: u32 = 165;
+const KEY_MEDIASTOP: u32 = 166;
```
Comment (main--) [inline windows-gaming-driver/src/driver/hotkeys/linux.rs]: correct names would be

```
#define KEY_NEXTSONG		163
#define KEY_PLAYPAUSE		164
#define KEY_PREVIOUSSONG	165
#define KEY_STOPCD		166
```

*shrug*

Comment (main--) [inline windows-gaming-driver/src/driver/hotkeys/linux.rs]: @oberien correct names would be

```
#define KEY_NEXTSONG		163
#define KEY_PLAYPAUSE		164
#define KEY_PREVIOUSSONG	165
#define KEY_STOPCD		166
```

*shrug*

File: windows-gaming-driver/src/driver/hotkeys/mod.rs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        KeyBinding { modifiers, no_repeat, key }
+    }
+
+    pub fn matches(&self, modifiers: &[Modifier], key: Key) -> bool {
+        key == self.key && self.modifiers.iter().all(|x| modifiers.contains(x))
+    }
+
+    pub fn to_windows(&self) -> (u32, u32) {
+        let base = if self.no_repeat { NOREPEAT } else { 0 };
+        (self.modifiers.iter().fold(base, |sum, &x| sum | (x as u32)), self.key as u32)
```
Comment (oberien) [inline windows-gaming-driver/src/driver/hotkeys/mod.rs]: Please add curly parenthesis around `sum | (x as u32)` to distinguish it from the closure-arguments.

**Review by moritzuehling — DISMISSED** [review]

File: windows-gaming-driver/src/driver/libinput.rs [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            Event::Keyboard(k) => println!("key {} {:?}", k.key(), k.key_state()),
+            Event::Pointer(m) => unreachable!(),
+            _ => (),
+        }
+        Ok(())
+    }).then(|_| Ok(()));
+
+    core.run(input_listener.join(input_handler)).unwrap();
+}
+ */
```
Comment (oberien) [inline windows-gaming-driver/src/driver/libinput.rs:254]: Still needed?

File: windows-gaming-driver/src/driver/mod.rs [hunk 13]
```diff
@@ -71,13 +80,99 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
...
+    let ref input_ref = *input;
+    let input_listener = libinput::InputListener(input_ref);
+    let hotkey_bindings: Vec<_> = cfg.machine.hotkeys.iter().map(|x| x.key.clone()).collect();
+    let mut keyboard_state = KeyboardState::new(&hotkey_bindings);
+    let input_handler = input_events.filter_map(|event| {
+        use self::monitor::*;
+        use std::iter;
+        use input::event::{Event, KeyboardEvent, PointerEvent};
+        use input::event::pointer::{Axis, ButtonState};
+        use input::event::keyboard::{KeyState, KeyboardEventTrait};
```
Comment (oberien) [inline windows-gaming-driver/src/driver/mod.rs]: welp

File: windows-gaming-driver/src/driver/qemu.rs [hunk 14]
```diff
@@ -138,16 +144,30 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
...
+                }
+            }
+        }
+
+        if bus == UsbBus::Xhci {
+            // add lighthouse usb-mouse
+            let port = i;
+            qemu.args(&["-device", &format!("usb-mouse,bus=xhci{}.0,port={}",
+                                            port / usable_ports, (port % usable_ports) + 1)]);
+            debug!("usb-mouse at xhci{}.0p{}", port / usable_ports, (port % usable_ports) + 1);
```
Comment (oberien) [inline windows-gaming-driver/src/driver/qemu.rs:170]: If it's stupid, but it works, it ain't stupid.

File: windows-gaming-driver/src/driver/mod.rs [hunk 15]
```diff
@@ -71,13 +80,99 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
...
+                    },
+                    None => return None
+                }
+            }
+            event => {
+                info!("Unhandled input event {:?}", event);
+                return None;
+            }
+        })
+    }).forward(monitor_sender.sink_map_err(|_| ())).then(|_| Ok(()));
```
Comment (oberien) [inline windows-gaming-driver/src/driver/mod.rs]: Could you please try to inline all mods into this file?

**Review by oberien — CHANGES_REQUESTED** [review]

File: windows-gaming-driver/src/driver/hotkeys/linux.rs (hunk 10 above)
Comment (oberien) [reply windows-gaming-driver/src/driver/hotkeys/linux.rs]: Done

**Review by oberien — APPROVED** [review]

**Pull request merged**